log4j-slf4j18-impl/src/main/java/org/apache/logging/slf4j/Log4jLoggerFactory.java

```diff
@@ -47,8 +47,12 @@ protected Logger newLogger(final String name, final LoggerContext context) {
 
     @Override
     protected LoggerContext getContext() {
-        final Class<?> anchor = StackLocatorUtil.getCallerClass(FQCN, PACKAGE);
-        return anchor == null ? LogManager.getContext() : getContext(StackLocatorUtil.getCallerClass(anchor));
+        final Class<?> anchor = LogManager.getFactory().isClassLoaderDependent()
+                ? StackLocatorUtil.getCallerClass(FQCN, PACKAGE)
+                : null;
+        return anchor == null
+                ? LogManager.getContext()
+                : getContext(StackLocatorUtil.getCallerClass(anchor));
```

Comment:
In the case where the context selector is classloader dependent, I imagine we could merge the `anchor = StackLocatorUtil.getCallerClass(FQCN, PACKAGE)` and `StackLocatorUtil.getCallerClass(anchor)` calls into a single sweep, potentially cutting the work in half.

Reply:
You would have to make a new method that gets the caller of the caller.

Reply:
That's what I had in mind. We could provide an overload which takes an additional integer parameter of additional frames to skip.
I'd prefer to implement that optimization separately from this change, and also implement another ContextSelector equivalent to `BasicContextSelector` for fully asynchronous logging so folks don't have to write their own to avoid classloader lookups.

Reply:
I can iterate on this PR with each component in an isolated commit, that may be easiest to review. Let me know if you think the current approach adding `boolean isClassLoaderDependent()` is reasonable.

Reply:
Yes, the optimization should be done separately. I'm not exactly sure what you mean by iterating on this PR. I would suggest merging this one and then do the other work as follow-up items later. So long as what you merge works we should be fine.